File: frontend/src/types/habitat.ts
export interface ScenarioHabitat {
  id: number;
  nom: string;
  description?: string;
  budget_estime?: number;
  surface_finale_m2?: number;
  nb_chambres?: number;
  score_global?: number;
  avantages: string[];
  inconvenients: string[];
  notes?: string;
  statut: string;
}

export interface CritereScenarioHabitat {
  id: number;
  scenario_id: number;
  nom: string;
  poids?: number;
  note?: number;
  commentaire?: string;
  score_global?: number;
}

export interface CritereImmoHabitat {
  id: number;
  nom: string;
  departements: string[];
  villes: string[];
  rayon_km: number;
  budget_min?: number;
  budget_max?: number;
  seuil_alerte?: number;
  actif: boolean;
}

export interface AnnonceHabitat {
  id: number;
  source: string;
  url_source: string;
  titre?: string;
  prix?: number;
  surface_m2?: number;
  surface_terrain_m2?: number;
  nb_pieces?: number;
  ville?: string;
  code_postal?: string;
  statut: string;
  score_pertinence?: number;
  prix_m2_secteur?: number;
  ecart_prix_pct?: number;
}

export interface PlanHabitat {
  id: number;
  scenario_id?: number;
  nom: string;
  type_plan: string;
  image_importee_url?: string;
  surface_totale_m2?: number;
  budget_estime?: number;
  version: number;
  suggestions_ia?: SuggestionPieceHabitat[];
}

export interface PieceHabitat {
  id: number;
  plan_id: number;
  nom: string;
  type_piece?: string;
  surface_m2?: number;
}

export interface ProjetDecoHabitat {
  id: number;
  piece_id?: number;
  nom_piece: string;
  style?: string;
  palette_couleurs?: string[];
  inspirations?: string[];
  budget_prevu?: number;
  budget_depense?: number;
  statut: string;
}

export interface ZoneJardinHabitat {
  id: number;
  plan_id: number;
  nom: string;
  type_zone?: string;
  surface_m2?: number;
  position_x?: number;
  position_y?: number;
  largeur?: number;
  longueur?: number;
  donnees_canvas?: import("@/types/commun").ObjetDonnees;
  plantes?: string[];
  amenagements?: string[];
  budget_estime?: number;
}

export interface HabitatHub {
  scenarios: number;
  annonces: number;
  plans: number;
  projets_deco: number;
  zones_jardin: number;
  alertes: number;
  annonces_a_traiter: number;
  budget_deco_total: number;
  budget_deco_depense: number;
}

export interface AlerteHabitat {
  id?: number;
  titre?: string;
  source: string;
  ville?: string;
  score?: number;
  prix?: number;
  ecart_prix_pct?: number;
  url_source: string;
  statut?: string;
}

export interface PointCarteHabitat {
  ville?: string;
  code_postal?: string;
  nb_annonces: number;
  score_max: number;
  prix_min?: number;
  prix_max?: number;
  latitude?: number;
  longitude?: number;
}

export interface StatSourceVeilleHabitat {
  source: string;
  url: string;
  annonces: number;
  alertes: number;
}

export interface ResultatSynchronisationVeilleHabitat {
  criteres: number;
  annonces_creees: number;
  annonces_mises_a_jour: number;
  alertes: number;
  sources: string[];
  stats_sources: StatSourceVeilleHabitat[];
}

export interface SuggestionPieceHabitat {
  nom: string;
  type_piece: string;
  surface_m2?: number;
  actions: string[];
}

export interface AnalysePlanHabitat {
  resume: string;
  points_forts: string[];
  risques: string[];
  budget_estime?: number;
  circulation_note?: number;
  suggestions_pieces: SuggestionPieceHabitat[];
  prompt_image?: string | null;
}

export interface HistoriquePlanHabitat {
  id: number;
  prompt_utilisateur: string;
  analyse_ia: AnalysePlanHabitat;
  image_generee_url?: string | null;
  acceptee?: boolean | null;
}

export interface ResultatImageHabitat {
  statut: string;
  modele: string;
  prompt: string;
  image_base64?: string | null;
  message?: string | null;
}

export interface ResultatAnalysePlanHabitat {
  plan_id: number;
  historique_id: number;
  analyse: AnalysePlanHabitat;
  image?: ResultatImageHabitat | null;
}

export interface ConceptDecoHabitat {
  resume: string;
  palette_couleurs: string[];
  materiaux: string[];
  achats_prioritaires: string[];
  prompt_image?: string | null;
}

export interface ResultatConceptDecoHabitat {
  projet_id: number;
  concept: ConceptDecoHabitat;
  image?: ResultatImageHabitat | null;
}

export interface ResumeJardinHabitat {
  zones: number;
  surface_totale_m2: number;
  budget_estime: number;
  types: string[];
}

export interface HistoriqueMarcheHabitatPoint {
  mois: string;
  transactions: number;
  prix_m2_median?: number | null;
  prix_m2_moyen?: number | null;
  valeur_moyenne?: number | null;
}

export interface RepartitionTypeMarcheHabitat {
  type_local: string;
  transactions: number;
  prix_m2_median?: number | null;
}

export interface TransactionMarcheHabitat {
  date_mutation: string;
  valeur_fonciere?: number | null;
  surface_m2?: number | null;
  prix_m2?: number | null;
  type_local?: string | null;
  nb_pieces?: number | null;
  code_postal?: string | null;
  commune?: string | null;
  adresse?: string | null;
}

export interface ResultatMarcheHabitat {
  source: {
    dataset_id: string;
    resource_id?: string | null;
    resource_title?: string | null;
  };
  query: {
    departement?: string | null;
    commune?: string | null;
    code_postal?: string | null;
    type_local?: string | null;
    nb_pieces_min?: number | null;
    surface_min_m2?: number | null;
  };
  resume: {
    nb_transactions: number;
    prix_m2_median?: number | null;
    prix_m2_moyen?: number | null;
    valeur_mediane?: number | null;
    surface_mediane?: number | null;
    dernier_mois?: HistoriqueMarcheHabitatPoint | null;
  };
  historique: HistoriqueMarcheHabitatPoint[];
  repartition_types: RepartitionTypeMarcheHabitat[];
  transactions: TransactionMarcheHabitat[];
}

File: src/api/schemas/habitat.py
"""Schémas Pydantic pour le module Habitat."""

from __future__ import annotations

from decimal import Decimal

from pydantic import BaseModel, Field


class ScenarioHabitatCreate(BaseModel):
    nom: str = Field(..., min_length=1, max_length=200)
    description: str | None = None
    budget_estime: Decimal | None = Field(None, ge=0)
    surface_finale_m2: Decimal | None = Field(None, ge=0)
    nb_chambres: int | None = Field(None, ge=0, le=20)
    avantages: list[str] = Field(default_factory=list)
    inconvenients: list[str] = Field(default_factory=list)
    notes: str | None = None
    statut: str = Field(default="brouillon", max_length=50)


class ScenarioHabitatPatch(BaseModel):
    nom: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = None
    budget_estime: Decimal | None = Field(None, ge=0)
    surface_finale_m2: Decimal | None = Field(None, ge=0)
    nb_chambres: int | None = Field(None, ge=0, le=20)
    avantages: list[str] | None = None
    inconvenients: list[str] | None = None
    notes: str | None = None
    statut: str | None = Field(None, max_length=50)


class CritereScenarioCreate(BaseModel):
    nom: str = Field(..., min_length=1, max_length=200)
    poids: Decimal = Field(default=Decimal("1.00"), ge=0, le=10)
    note: Decimal | None = Field(None, ge=0, le=10)
    commentaire: str | None = None


class CritereImmoCreate(BaseModel):
    nom: str = Field(default="Recherche principale", max_length=200)
    departements: list[str] = Field(default_factory=list)
    villes: list[str] = Field(default_factory=list)
    rayon_km: int = Field(default=10, ge=1, le=100)
    budget_min: Decimal | None = Field(None, ge=0)
    budget_max: Decimal | None = Field(None, ge=0)
    surface_min_m2: Decimal | None = Field(None, ge=0)
    surface_terrain_min_m2: Decimal | None = Field(None, ge=0)
    nb_pieces_min: int | None = Field(None, ge=1)
    nb_chambres_min: int | None = Field(None, ge=1)
    type_bien: str | None = Field(None, max_length=50)
    criteres_supplementaires: dict = Field(default_factory=dict)
    seuil_alerte: Decimal = Field(default=Decimal("0.70"), ge=0, le=1)
    actif: bool = True


class AnnonceHabitatCreate(BaseModel):
    critere_id: int | None = None
    source: str = Field(..., min_length=2, max_length=100)
    url_source: str = Field(..., min_length=8, max_length=500)
    titre: str | None = Field(None, max_length=500)
    prix: Decimal | None = Field(None, ge=0)
    surface_m2: Decimal | None = Field(None, ge=0)
    surface_terrain_m2: Decimal | None = Field(None, ge=0)
    nb_pieces: int | None = Field(None, ge=0)
    ville: str | None = Field(None, max_length=200)
    code_postal: str | None = Field(None, max_length=10)
    departement: str | None = Field(None, max_length=3)
    photos: list[str] = Field(default_factory=list)
    description_brute: str | None = None
    score_pertinence: Decimal | None = Field(None, ge=0, le=1)
    statut: str = Field(default="nouveau", max_length=50)
    prix_m2_secteur: Decimal | None = Field(None, ge=0)
    ecart_prix_pct: Decimal | None = None
    hash_dedup: str | None = Field(None, max_length=64)
    notes: str | None = None


class PlanHabitatCreate(BaseModel):
    scenario_id: int | None = None
    nom: str = Field(..., min_length=1, max_length=200)
    type_plan: str = Field(..., min_length=3, max_length=50)
    image_importee_url: str | None = Field(None, max_length=500)
    donnees_pieces: dict = Field(default_factory=dict)
    contraintes: dict = Field(default_factory=dict)
    surface_totale_m2: Decimal | None = Field(None, ge=0)
    budget_estime: Decimal | None = Field(None, ge=0)
    notes: str | None = None


class PlanHabitatAnalyseCreate(BaseModel):
    prompt_utilisateur: str | None = Field(None, max_length=2000)
    generer_image: bool = False


class PlanHabitatPiece3DConfig(BaseModel):
    id: int = Field(..., ge=1)
    x: float
    z: float
    width: float = Field(..., ge=1.8, le=12)
    depth: float = Field(..., ge=1.8, le=12)
    nom: str | None = Field(None, min_length=1, max_length=200)
    type_piece: str | None = Field(None, max_length=50)


class PlanHabitatConfiguration3D(BaseModel):
    layout_edition: list[PlanHabitatPiece3DConfig] = Field(default_factory=list)
    palette_par_type: dict[str, str] = Field(default_factory=dict)


class PlanHabitatVariante3D(BaseModel):
    id: str | None = Field(None, min_length=1, max_length=120)
    nom: str = Field(..., min_length=1, max_length=120)
    source: str = Field(default="manuel", max_length=50)
    configuration: PlanHabitatConfiguration3D


class PlanHabitatConfiguration3DUpdate(BaseModel):
    configuration_courante: PlanHabitatConfiguration3D = Field(default_factory=PlanHabitatConfiguration3D)
    variantes: list[PlanHabitatVariante3D] = Field(default_factory=list)
    variante_active_id: str | None = Field(None, min_length=1, max_length=120)


class PieceHabitatCreate(BaseModel):
    nom: str = Field(..., min_length=1, max_length=200)
    type_piece: str | None = Field(None, max_length=50)
    surface_m2: Decimal | None = Field(None, ge=0)
    position_x: Decimal | None = None
    position_y: Decimal | None = None
    largeur: Decimal | None = Field(None, ge=0)
    longueur: Decimal | None = Field(None, ge=0)
    hauteur_plafond: Decimal | None = Field(None, ge=0)
    couleur_mur: str | None = Field(None, max_length=7)
    sol_type: str | None = Field(None, max_length=50)
    meubles: list = Field(default_factory=list)
    notes: str | None = None


class ProjetDecoHabitatCreate(BaseModel):
    piece_id: int | None = None
    nom_piece: str = Field(..., min_length=1, max_length=200)
    style: str | None = Field(None, max_length=100)
    palette_couleurs: list = Field(default_factory=list)
    inspirations: list = Field(default_factory=list)
    budget_prevu: Decimal | None = Field(None, ge=0)
    notes: str | None = None


class ProjetDecoSuggestionCreate(BaseModel):
    brief: str | None = Field(None, max_length=2000)
    generer_image: bool = False


class ProjetDecoDepenseCreate(BaseModel):
    montant: Decimal = Field(..., gt=0)
    fournisseur: str | None = Field(None, max_length=200)
    note: str | None = None
    categorie_depense: str = Field(default="travaux", max_length=50)


class ZoneJardinHabitatCreate(BaseModel):
    plan_id: int
    nom: str = Field(..., min_length=1, max_length=200)
    type_zone: str | None = Field(None, max_length=100)
    surface_m2: Decimal | None = Field(None, ge=0)
    altitude_relative: Decimal | None = None
    position_x: Decimal | None = None
    position_y: Decimal | None = None
    largeur: Decimal | None = Field(None, ge=0)
    longueur: Decimal | None = Field(None, ge=0)
    donnees_canvas: dict = Field(default_factory=dict)
    plantes: list = Field(default_factory=list)
    amenagements: list = Field(default_factory=list)
    budget_estime: Decimal | None = Field(None, ge=0)
    notes: str | None = None


class ZoneJardinHabitatPatch(BaseModel):
    nom: str | None = Field(None, min_length=1, max_length=200)
    type_zone: str | None = Field(None, max_length=100)
    surface_m2: Decimal | None = Field(None, ge=0)
    altitude_relative: Decimal | None = None
    position_x: Decimal | None = None
    position_y: Decimal | None = None
    largeur: Decimal | None = Field(None, ge=0)
    longueur: Decimal | None = Field(None, ge=0)
    donnees_canvas: dict | None = None
    plantes: list | None = None
    amenagements: list | None = None
    budget_estime: Decimal | None = Field(None, ge=0)
    notes: str | None = None


class SynchronisationVeilleHabitatCreate(BaseModel):
    critere_id: int | None = None
    limite_par_source: int = Field(default=12, ge=1, le=50)
    sources: list[str] | None = None
    envoyer_alertes: bool = True


class GenerationImageHabitatCreate(BaseModel):
    prompt: str = Field(..., min_length=10, max_length=2000)
    negative_prompt: str | None = Field(None, max_length=1000)
